concordium-node/src/consensus_ffi/ffi.rs

```diff
@@ -2990,62 +3002,65 @@ pub extern "C" fn on_finalization_message_catchup_out(
-
-            let payload = slice::from_raw_parts($msg as *const u8, $msg_length as usize);
-            let mut full_payload = Vec::with_capacity(5 + payload.len());
-            (msg_variant as u8).serial(&mut full_payload);
-            ($genesis_index as u32).serial(&mut full_payload);
-            full_payload.write_all(&payload).unwrap(); // infallible
-            let full_payload = Arc::from(full_payload);
-
-            let msg = ConsensusMessage::new(
-                MessageType::Outbound($target),
-                msg_variant,
-                full_payload,
-                vec![],
-                $omit_status,
-            );
-
-            match CALLBACK_QUEUE.send_out_blocking_msg(msg) {
-                Ok(_) => trace!("Queueing a {} of {} bytes", msg_variant, $msg_length),
-                Err(e) => error!("Couldn't queue a {} properly: {}", msg_variant, e),
-            };
-        }
+/// Helper function for deserializing and enqueing consensus messages in
+/// CALLBACK_QUEUE.
+fn sending_callback(
+    target: Option<RemotePeerId>,
+    callback_type: CallbackType,
+    genesis_index: u32,
+    msg: *const u8,
+    msg_length: usize,
+    omit_status: Option<PeerStatus>,
+) {
+    let msg_variant = match callback_type {
+        CallbackType::Block => PacketType::Block,
+        CallbackType::FinalizationMessage => PacketType::FinalizationMessage,
+        CallbackType::FinalizationRecord => PacketType::FinalizationRecord,
+        CallbackType::CatchUpStatus => PacketType::CatchUpStatus,
+    };
+
+    let payload = unsafe { slice::from_raw_parts(msg, msg_length) };
+    let mut full_payload = Vec::with_capacity(5 + payload.len());
+    (msg_variant as u8).serial(&mut full_payload);
+    (genesis_index).serial(&mut full_payload);
+    full_payload.write_all(payload).unwrap(); // infallible
+    let full_payload = Arc::from(full_payload);
+
+    let msg = ConsensusMessage::new(
+        MessageType::Outbound(target),
+        msg_variant,
+        full_payload,
+        vec![],
+        omit_status,
+    );
+
+    match CALLBACK_QUEUE.send_out_blocking_msg(msg) {
+        Ok(_) => trace!("Queueing a {} of {} bytes", msg_variant, msg_length),
+        Err(e) => error!("Couldn't queue a {} properly: {}", msg_variant, e),
     };
 }
 
-pub extern "C" fn broadcast_callback(
+pub unsafe extern "C" fn broadcast_callback(
+    context_ptr: *const BroadcastCallbackContext,
     msg_type: i64,
     genesis_index: u32,
     msg: *const u8,
     msg_length: i64,
 ) {
     trace!("Broadcast callback hit - queueing message");
-    sending_callback!(None, msg_type, genesis_index, msg, msg_length, None);
+    let callback_type = match CallbackType::try_from(msg_type as u8) {
+        Ok(ct) => ct,
+        Err(e) => {
+            error!("{}", e);
+            return;
+        }
+    };
+
+    if let CallbackType::Block = callback_type {
+        let context = unsafe { context_ptr.read() };
+        context.metric_baked_blocks.inc();
+    }
```

Comment:
Why not do this inside the sending_callback? The type is deserialized there as well.